File: server/src/modules/gameSession/parts/collision.ts
import {
  ARENA_HEIGHT,
  BALL_SIZE,
  BALL_SPEED_INCREMENT,
  BALL_MAX_SPEED,
  PADDLE_WIDTH,
} from "../../../config";
import type { BallState } from "../../../shared";

const WALL_THICKNESS = 5;
const HALF_H = ARENA_HEIGHT / 2;
const HALF_BALL = BALL_SIZE / 2;

export interface CollisionResult {
  ballHitWall: boolean;
  ballHitPaddle: boolean;
}

export function checkWallCollisions(ball: BallState): boolean {
  let hit = false;

  if (ball.y + HALF_BALL > HALF_H - WALL_THICKNESS) {
    ball.vy = -Math.abs(ball.vy);
    ball.y = HALF_H - WALL_THICKNESS - HALF_BALL;
    hit = true;
  } else if (ball.y - HALF_BALL < -HALF_H + WALL_THICKNESS) {
    ball.vy = Math.abs(ball.vy);
    ball.y = -HALF_H + WALL_THICKNESS + HALF_BALL;
    hit = true;
  }

  return hit;
}

export function checkPaddleCollision(
  ball: BallState,
  paddleY: number,
  paddleX: number,
  paddleHeight: number,
): boolean {
  if (!aabbCollision(ball.x, ball.y, BALL_SIZE, paddleX, paddleY, PADDLE_WIDTH, paddleHeight)) {
    return false;
  }

  const hitOffset = (ball.y - paddleY) / (paddleHeight / 2);
  const isRightPaddle = paddleX > 0;

  applyPaddleBounce(ball, hitOffset, isRightPaddle);

  // Push ball out of paddle
  if (paddleX < 0) {
    ball.x = paddleX + PADDLE_WIDTH / 2 + HALF_BALL + 1;
  } else {
    ball.x = paddleX - PADDLE_WIDTH / 2 - HALF_BALL - 1;
  }

  return true;
}

function applyPaddleBounce(ball: BallState, hitOffset: number, isRightPaddle: boolean): void {
  const currentSpeed = Math.sqrt(ball.vx * ball.vx + ball.vy * ball.vy);
  const speed = Math.min(currentSpeed, BALL_MAX_SPEED) + BALL_SPEED_INCREMENT;
  const angle = Math.max(-1, Math.min(1, hitOffset)) * (Math.PI / 4);

  const dirX = isRightPaddle ? -1 : 1;

  ball.vx = dirX * speed * Math.cos(angle);
  ball.vy = speed * Math.sin(angle);
}

function aabbCollision(
  ax: number, ay: number, aSize: number,
  bx: number, by: number, bWidth: number, bHeight: number,
): boolean {
  const aHalf = aSize / 2;
  const bHalfW = bWidth / 2;
  const bHalfH = bHeight / 2;

  return (
    ax - aHalf < bx + bHalfW &&
    ax + aHalf > bx - bHalfW &&
    ay - aHalf < by + bHalfH &&
    ay + aHalf > by - bHalfH
  );
}

File: client/src/modules/game/parts/collision.rs
use bevy::prelude::*;
use pong_shared::config::*;

use crate::shared::components::*;
use crate::shared::events::CollisionEvent;

use super::physics;

pub fn check_collisions(
    mut ball_query: Query<(&mut Transform, &mut Velocity), With<Ball>>,
    paddle_query: Query<(&Transform, &PaddleStats), (With<Paddle>, Without<Ball>)>,
    mut collision_events: MessageWriter<CollisionEvent>,
) {
    let half_h = ARENA_HEIGHT / 2.0;
    let wall_thickness = 5.0;

    for (mut ball_transform, mut ball_velocity) in &mut ball_query {
        let ball_pos = ball_transform.translation;

        // Wall collisions (top/bottom)
        if ball_pos.y + BALL_SIZE / 2.0 > half_h - wall_thickness {
            ball_velocity.0.y = -ball_velocity.0.y.abs();
            ball_transform.translation.y = half_h - wall_thickness - BALL_SIZE / 2.0;
            collision_events.write(CollisionEvent);
        } else if ball_pos.y - BALL_SIZE / 2.0 < -half_h + wall_thickness {
            ball_velocity.0.y = ball_velocity.0.y.abs();
            ball_transform.translation.y = -half_h + wall_thickness + BALL_SIZE / 2.0;
            collision_events.write(CollisionEvent);
        }

        // Paddle collisions
        for (paddle_transform, paddle_stats) in &paddle_query {
            let paddle_pos = paddle_transform.translation;
            let paddle_h = paddle_stats.height;

            if aabb_collision(ball_pos, BALL_SIZE, paddle_pos, PADDLE_WIDTH, paddle_h) {
                let hit_offset = (ball_pos.y - paddle_pos.y) / (paddle_h / 2.0);

                physics::apply_paddle_bounce(&mut ball_velocity, hit_offset, paddle_pos.x > 0.0);

                // Push ball out of paddle
                if paddle_pos.x < 0.0 {
                    ball_transform.translation.x =
                        paddle_pos.x + PADDLE_WIDTH / 2.0 + BALL_SIZE / 2.0 + 1.0;
                } else {
                    ball_transform.translation.x =
                        paddle_pos.x - PADDLE_WIDTH / 2.0 - BALL_SIZE / 2.0 - 1.0;
                }

                collision_events.write(CollisionEvent);
            }
        }
    }
}

fn aabb_collision(
    a_pos: Vec3,
    a_size: f32,
    b_pos: Vec3,
    b_width: f32,
    b_height: f32,
) -> bool {
    let a_half = a_size / 2.0;
    let b_half_w = b_width / 2.0;
    let b_half_h = b_height / 2.0;

    a_pos.x - a_half < b_pos.x + b_half_w
        && a_pos.x + a_half > b_pos.x - b_half_w
        && a_pos.y - a_half < b_pos.y + b_half_h
        && a_pos.y + a_half > b_pos.y - b_half_h
}
